[AutoDiff] TF-123: fix reabstraction to opaque

marcrasi opened the pull request:
This fixes TF-123 by adding a new abstraction pattern that properly abstracts the derivative functions in a `differentiable_function` bundle that is being abstracted to `AP::Opaque`.

Even though this technically fixes TF-123 (all its testcases compile and run successfully), many simple operations involving opaquely differentiable functions are still broken due to TF-1122. I believe that TF-1122 is a distinct problem, so it makes sense to fix it later in a separate PR.

This also adds tests for reabstraction:
- The last 3 tests in `reabstraction_e2e.swift` are the ones that have been fixed by this PR.
- The last 3 tests include some commented-out code that currently segfaults at runtime due to TF-1122.
- All the tests in `reabstraction_e2e.swift` except for the last 3 already pass before this PR. I just used this PR as an opportunity to test reabstraction more thoroughly.

The comment that I added in `AbstractionPattern.h` describes the solution in more detail:

```
    /// A Swift function whose parameters and results are opaque. This is
    /// like `AP::Type<T>((T) -> T)`, except that the number of parameters is
    /// unspecified.
    ///
    /// This is used to construct the abstraction pattern for the
    /// derivative function of a function with opaque abstraction pattern. See
    /// `OpaqueDerivativeFunction`.
    OpaqueFunction,
    /// A Swift function whose parameters are opaque and whose result is the
    /// tuple abstraction pattern `(AP::Opaque, AP::OpaqueFunction)`.
    ///
    /// Purpose. When we reabstract `differentiable_function` using the
    /// `AP::Opaque` pattern, we use `AP::Opaque` to reabstract the original
    /// function in the bundle and we use `AP::OpaqueDerivativeFunction` to
    /// reabstract the derivative functions in the bundle. This preserves the
    /// `differentiable_function` invariant that the derivative type
    /// (`SILFunctionType::getAutoDiffDerivativeFunctionType()`) of the original
    /// function is equal to the type of the derivative function. For example:
    ///
    ///   differentiable_function
    ///     [parameters 0]
    ///     %0 : $@callee_guaranteed (Float) -> Float
    ///     with_derivative {
    ///       %1 : $@callee_guaranteed (Float) -> (
    ///         Float,
    ///         @owned @callee_guaranteed (Float) -> Float
    ///       ),
    ///       %2 : $@callee_guaranteed (Float) -> (
    ///         Float,
    ///         @owned @callee_guaranteed (Float) -> Float
    ///       )
    ///     }
    ///
    /// The invariant-respecting abstraction of this bundle to `AP::Opaque` is:
    ///
    ///   differentiable_function
    ///     [parameters 0]
    ///     %3 : $@callee_guaranteed (@in_guaranteed Float) -> @out Float
    ///     with_derivative {
    ///       %4 : $@callee_guaranteed (@in_guaranteed Float) -> (
    ///         @out Float,
    ///         @owned @callee_guaranteed (@in_guaranteed Float) -> @out Float
    ///       ),
    ///       %5 : $@callee_guaranteed (@in_guaranteed Float) -> (
    ///         @out Float,
    ///         @owned @callee_guaranteed (@in_guaranteed Float) -> @out Float
    ///       )
    ///     }
    ///
    /// In particular:
    ///
    /// - The reabstraction %0 => %3 uses pattern `AP::Opaque`.
    /// - The reabstraction %1 => %4 uses pattern
    ///   `AP::OpaqueDerivativeFunction`, which maximally abstracts all the
    ///   parameters, and abstracts the result as the tuple
    ///   `(AP::Opaque, AP::OpaqueFunction)`.
    /// - The reabstraction %2 => %5 similarly uses pattern
    ///   `AP::OpaqueDerivativeFunction`.
    OpaqueDerivativeFunction,
```

Commit:
[AutoDiff] TF-123: fix reabstraction to opaque

dan-zheng commented on lib/SIL/AbstractionPattern.cpp:
```diff
@@ -869,3 +910,26 @@ const {
+AbstractionPattern AbstractionPattern::getAutoDiffDerivativeFunctionType(
+    IndexSubset *indices, unsigned resultIndex,
+    AutoDiffDerivativeFunctionKind kind, LookupConformanceFn lookupConformance,
+    GenericSignature whereClauseGenericSignature, bool makeSelfParamFirst) {
```
```suggestion
    GenericSignature derivativeGenericSignature, bool makeSelfParamFirst) {
```

marcrasi replied:
done

dan-zheng commented on include/swift/SIL/AbstractionPattern.h:
```diff
@@ -1020,6 +1126,19 @@ class AbstractionPattern {
+      IndexSubset *indices, unsigned resultIndex,
+      AutoDiffDerivativeFunctionKind kind,
+      LookupConformanceFn lookupConformance,
+      GenericSignature whereClauseGenericSignature = GenericSignature(),
```
Minor: the `whereClauseGenericSignature` argument of `AnyFunctionType::getAutoDiffDerivativeFunctionType` has been renamed to `derivativeGenericSignature` [on `master`](https://github.com/apple/swift/blob/fa1544974b67bcfd2c1cba1023cfb3a381ce7133/include/swift/AST/Types.h#L3260):
```suggestion
      GenericSignature derivativeGenericSignature = GenericSignature(),
```
The renaming will soon propagate to `tensorflow` branch.

marcrasi replied:
done

dan-zheng commented on test/AutoDiff/downstream/reabstraction_e2e.swift:
```diff
@@ -0,0 +1,178 @@
+  let inner: @differentiable (Float) -> Float = { 7 * $0 * $0 }
+  let transformed = id(inner)
+
+  // TODO(TF-1122): Actually using `transformed` causes a segfault at runtime.
```
How about adding a dedicated `not --crash` negative test for TF-1122? In addition to commenting out the crashers in this test.

marcrasi replied:
done

dan-zheng commented on include/swift/SIL/AbstractionPattern.h:
```diff
@@ -238,7 +298,8 @@ class AbstractionPattern {
 
-  unsigned TheKind : 32 - NumOtherDataBits;
+  // SWIFT_ENABLE_TENSORFLOW
+  unsigned TheKind : 33 - NumOtherDataBits;
```
Note: I wonder if this change is significant e.g. for upstreaming. Maybe not.

dan-zheng reviewed (approved):
Great solution to a longstanding issue! Nice documentation.

futurejones commented:
Can one of the admins verify this patch?

Commit:
Merge branch 'tensorflow' into tf-123

Commit:
address comments

marcrasi commented:
@swift-ci please test tensorflow

marcrasi commented:
@swift-ci please test tensorflow

Commit:
Merge branch 'tensorflow' into tf-123

Commit:
fix test

marcrasi commented:
@swift-ci please test tensorflow

marcrasi commented:
@swift-ci please test tensorflow

marcrasi commented:
@swift-ci please test tensorflow

marcrasi commented:
@swift-ci please test tensorflow

marcrasi commented:
@swift-ci please test tensorflow

marcrasi commented:
@swift-ci please test tensorflow

Commit:
fix test again

marcrasi commented:
@swift-ci please test tensorflow

marcrasi commented:
@swift-ci please test tensorflow

marcrasi commented:
@swift-ci please test tensorflow

marcrasi commented:
@swift-ci please test tensorflow

marcrasi commented:
@swift-ci please test tensorflow

marcrasi commented:
@swift-ci please test tensorflow macos

marcrasi commented:
@swift-ci please test tensorflow macos

marcrasi commented:
@swift-ci please test tensorflow macos

marcrasi commented:
@swift-ci please test tensorflow macos

marcrasi commented:
@swift-ci please test tensorflow macos

marcrasi commented:
@swift-ci please test tensorflow macos

marcrasi commented:
@swift-ci please test tensorflow macos

dan-zheng commented:
`@swift-ci Please test tensorflow` is still down, so I'll help test this PR on macOS manually.

dan-zheng committed:
Merge branch 'tensorflow' of github.com:apple/swift into tf-123

dan-zheng committed:
Minor edits.

- Add `SWIFT_ENABLE_TENSORFLOW END` markers.
- Rename `whereClauseGenericSignature` to `derivativeGenericSignature`.

dan-zheng committed:
Minor doc comment edits.

- `differentiable_function` -> `@differentiable` function value
- Minimize mentions of informal term "bundle".

dan-zheng commented:
I merged `apple/tensorflow` and pushed some minor edits.
`swift/utils/build-script --preset tensorflow_test` passes on macOS. Merging now!
